storage: don't fatal on inline DeleteRange that hits non-inline value

In a support case, we saw that an assertion like the following could be hit during inline (`timestamp == 0`) `DeleteRange` evaluation:
```
F230412 18:36:44.436420 52096636 kv/kvserver/pkg/kv/kvserver/replica_batch_updates.go:227 ⋮ [n1,s1,r774/1:‹/System/tsd/cr.store.{qu…-re…}›] 7351 trying to bump to 1681309128.304176583,1 <= ba.Timestamp: 1681324603.939164946,0
```

We suspect that this is due to corruption that allowed the DeleteRange to see an MVCC value (`timestamp != 0`).

Inline `DeleteRange` requests are incompatible with MVCC values, but the presence of these values should not lead to a crash. Instead, they should return an error:
https://github.com/cockroachdb/cockroach/blob/165c29a8ae338bd6948d886919d953fd167234bd/pkg/storage/mvcc.go#L1295-L1298

Our leading theory is that the corruption led to an unexpected `WriteTooOld` error because the `DeleteRange`'s search phase sets the `MVCCScanOptions.FailOnMoreRecent=true` scan option:
https://github.com/cockroachdb/cockroach/blob/165c29a8ae338bd6948d886919d953fd167234bd/pkg/storage/mvcc.go#L2179-L2181

We can demonstrate this with the following test:
```go
func TestMVCCDeleteRangeInlineEncountersMVCCVersion(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	ctx := context.Background()
	engine := NewDefaultInMemForTesting()
	defer engine.Close()

	err := MVCCPut(ctx, engine, nil, testKey1,
		hlc.Timestamp{WallTime: 1}, hlc.ClockTimestamp{}, value1, nil)
	require.NoError(t, err)

	_, _, _, err = MVCCDeleteRange(ctx, engine, nil, testKey1, testKey2,
		0, hlc.Timestamp{}, hlc.ClockTimestamp{}, nil, false)
	require.NoError(t, err)
}
```
The test currently fails with:
```
--- FAIL: TestMVCCDeleteRangeInlineEncountersMVCCVersion (0.02s)
    test_log_scope.go:161: test logs captured to: /tmp/_tmp/d1091de82354dd1e885605d51bd315be/logTestMVCCDeleteRangeInlineEncountersMVCCVersion2996635390
    test_log_scope.go:79: use -show-logs to present logs inline
    mvcc_test.go:1849:
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/storage/mvcc_test.go:1849
        	Error:      	Received unexpected error:
        	            	WriteTooOldError: write for key "/db1" at timestamp 0,0 too old; wrote at 0.000000001,1
        	Test:       	TestMVCCDeleteRangeInlineEncountersMVCCVersion
```

We should stop setting this flag for inline `DeleteRange` requests. We may also want to scan at MaxTimestamp instead of timestamp 0 so that non-inline values are returned to the mutation phase for the error check, instead of ignored by the search phase.

Jira issue: CRDB-26952

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -3000,6 +3000,31 @@ func MVCCDeleteRange(
 	txn *roachpb.Transaction,
 	returnKeys bool,
 ) ([]roachpb.Key, *roachpb.Span, int64, error) {
+	// Scan to find the keys to delete.
+	//
+	// For a versioned delete range, scan at the request timestamp and with the
+	// FailOnMoreRecent option set to true. Doing so returns all non-tombstoned
+	// keys at or below the request timestamp and throws a WriteTooOld error on
+	// any key (mvcc tombstone or otherwise) above the request timestamp. This is
+	// different from scanning at MaxTimestamp and deferring write-write conflict
+	// checking to mvccPutInternal below. That approach ignores mvcc tombstones
+	// above the request timestamp, which could lead to serializability anomalies
+	// (see #56458).
+	//
+	// For an inline delete range, scan at MaxTimestamp. Doing so is not needed to
+	// retrieve inline values, but it ensures that all non-inline values are also
+	// returned. It is incompatible to mix an inline delete range with mvcc
+	// versions, so we want to pass these incompatible keys to mvccPutInternal to
+	// detect the condition and return an error. We also scan with the
+	// FailOnMoreRecent set to false. This is not strictly necessary (nothing is
+	// more recent than MaxTimestamp), but it provides added protection against
+	// the scan returning a WriteTooOld error.
+	scanTs := timestamp
+	failOnMoreRecent := true
+	if timestamp.IsEmpty() /* inline */ {
+		scanTs = hlc.MaxTimestamp
+		failOnMoreRecent = false
+	}
 	// In order for this operation to be idempotent when run transactionally, we
 	// need to perform the initial scan at the previous sequence number so that
 	// we don't see the result from equal or later sequences.
@@ -3009,8 +3034,8 @@ func MVCCDeleteRange(
 		prevSeqTxn.Sequence--
 		scanTxn = prevSeqTxn
 	}
-	res, err := MVCCScan(ctx, rw, key, endKey, timestamp, MVCCScanOptions{
-		FailOnMoreRecent: true, Txn: scanTxn, MaxKeys: max,
+	res, err := MVCCScan(ctx, rw, key, endKey, scanTs, MVCCScanOptions{
+		FailOnMoreRecent: failOnMoreRecent, Txn: scanTxn, MaxKeys: max,
 	})
 	if err != nil {
 		return nil, nil, 0, err

--- a/pkg/storage/mvcc_test.go
+++ b/pkg/storage/mvcc_test.go
@@ -1780,11 +1780,10 @@ func TestMVCCDeleteRangeInline(t *testing.T) {
 	}
 
 	// Attempt to delete non-inline key at zero timestamp; should fail.
-	const writeTooOldErrString = "WriteTooOldError"
 	if _, _, _, err := MVCCDeleteRange(ctx, engine, nil, testKey6, keyMax,
 		1, hlc.Timestamp{Logical: 0}, hlc.ClockTimestamp{}, nil, true,
-	); !testutils.IsError(err, writeTooOldErrString) {
-		t.Fatalf("got error %v, expected error with text '%s'", err, writeTooOldErrString)
+	); !testutils.IsError(err, inlineMismatchErrString) {
+		t.Fatalf("got error %v, expected error with text '%s'", err, inlineMismatchErrString)
 	}
 
 	// Attempt to delete inline keys in a transaction; should fail.

--- a/pkg/storage/testdata/mvcc_histories/delete_range_inline
+++ b/pkg/storage/testdata/mvcc_histories/delete_range_inline
@@ -1,0 +1,51 @@
+# Mix inline and mvcc delete range requests with inline and mvcc key-values.
+
+run ok
+put k=inline-key v=inline-val ts=0
+----
+>> at end:
+meta: "inline-key"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline-val mergeTs=<nil> txnDidNotUpdateMeta=false
+
+run ok
+put k=mvcc-key v=mvcc-val ts=10
+----
+>> at end:
+meta: "inline-key"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline-val mergeTs=<nil> txnDidNotUpdateMeta=false
+data: "mvcc-key"/10.000000000,0 -> /BYTES/mvcc-val
+
+
+# Incompatible requests.
+
+run error
+del_range k=inline-key end=inline-key-end ts=20
+----
+>> at end:
+meta: "inline-key"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline-val mergeTs=<nil> txnDidNotUpdateMeta=false
+data: "mvcc-key"/10.000000000,0 -> /BYTES/mvcc-val
+error: (*withstack.withStack:) "inline-key"/0,0: put is inline=false, but existing value is inline=true
+
+run error
+del_range k=mvcc-key end=mvcc-key-end ts=0
+----
+>> at end:
+meta: "inline-key"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline-val mergeTs=<nil> txnDidNotUpdateMeta=false
+data: "mvcc-key"/10.000000000,0 -> /BYTES/mvcc-val
+error: (*withstack.withStack:) "mvcc-key"/0,0: put is inline=true, but existing value is inline=false
+
+
+# Compatible requests.
+
+run ok
+del_range k=inline-key end=inline-key-end ts=0
+----
+del_range: "inline-key"-"inline-key-end" -> deleted 1 key(s)
+>> at end:
+data: "mvcc-key"/10.000000000,0 -> /BYTES/mvcc-val
+
+run ok
+del_range k=mvcc-key end=mvcc-key-end ts=20
+----
+del_range: "mvcc-key"-"mvcc-key-end" -> deleted 1 key(s)
+>> at end:
+data: "mvcc-key"/20.000000000,0 -> /<empty>
+data: "mvcc-key"/10.000000000,0 -> /BYTES/mvcc-val

